File: nodejs/src/performance_client.js
/**
 * # High-Performance Node.js LLM Client
 * 
 * Addresses all performance failure modes:
 * - Connection pooling with keep-alive
 * - Single deadline propagated to all operations
 * - Proper streaming with backpressure
 * - Bounded concurrency with semaphores
 * - Memory-efficient buffer reuse
 */

const http = require('http');
const https = require('https');
const { URL } = require('url');
const { EventEmitter } = require('events');

class Semaphore {
    constructor(max) {
        this.max = max;
        this.current = 0;
        this.queue = [];
    }

    async acquire() {
        return new Promise((resolve) => {
            if (this.current < this.max) {
                this.current++;
                resolve();
            } else {
                this.queue.push(resolve);
            }
        });
    }

    release() {
        if (this.queue.length > 0) {
            const resolve = this.queue.shift();
            resolve();
        } else {
            this.current--;
        }
    }
}

class BufferPool {
    constructor(maxSize = 10) {
        this.pool = [];
        this.maxSize = maxSize;
    }

    get() {
        return this.pool.pop() || Buffer.allocUnsafe(8192);
    }

    return(buffer) {
        if (buffer.length <= 65536 && this.pool.length < this.maxSize) {
            buffer.fill(0); // Clear buffer
            this.pool.push(buffer);
        }
    }
}

class PerformanceClient {
    constructor(config = {}) {
        this.config = {
            baseUrl: 'http://localhost:3000',
            timeout: 30000,
            maxConcurrent: 32,
            keepAlive: 60000,
            retryAttempts: 3,
            retryBaseDelay: 100,
            maxRetryDelay: 5000,
            ...config
        };

        this.semaphore = new Semaphore(this.config.maxConcurrent);
        this.bufferPool = new BufferPool();
        
        // Connection pool - reuse connections
        this.agents = new Map();
        this.setupAgents();
    }

    setupAgents() {
        const parsedUrl = new URL(this.config.baseUrl);
        const isHttps = parsedUrl.protocol === 'https:';
        
        // HTTP/HTTPS agent with connection pooling
        const agentOptions = {
            keepAlive: true,
            keepAliveMsecs: this.config.keepAlive,
            maxSockets: this.config.maxConcurrent,
            maxFreeSockets: 5,
            timeout: this.config.timeout,
            scheduling: 'fifo'
        };

        if (isHttps) {
            this.agent = new https.Agent(agentOptions);
        } else {
            this.agent = new http.Agent(agentOptions);
        }
    }

    async chatCompletion(messages, deadline) {
        const permit = await this.semaphore.acquire();
        
        try {
            // Check deadline
            if (Date.now() > deadline) {
                throw new Error('Deadline exceeded');
            }

            const remainingTime = deadline - Date.now();
            const idempotencyKey = this.generateIdempotencyKey();
            
            const body = JSON.stringify({
                model: 'test-model',
                messages,
                max_tokens: 100
            });

            return await this.makeRequestWithRetries(body, remainingTime, idempotencyKey);
        } finally {
            this.semaphore.release();
        }
    }

    async streamChatCompletion(messages, deadline) {
        const permit = await this.semaphore.acquire();
        
        try {
            if (Date.now() > deadline) {
                throw new Error('Deadline exceeded');
            }

            const remainingTime = deadline - Date.now();
            const idempotencyKey = this.generateIdempotencyKey();
            
            const body = JSON.stringify({
                model: 'test-model',
                messages,
                max_tokens: 100,
                stream: true
            });

            return this.streamRequest(body, remainingTime, idempotencyKey);
        } finally {
            this.semaphore.release();
        }
    }

    async makeRequestWithRetries(body, remainingTime, idempotencyKey) {
        let attempt = 0;
        let lastError = null;
        const startTime = Date.now();

        while (attempt < this.config.retryAttempts) {
            attempt++;
            
            const elapsed = Date.now() - startTime;
            if (elapsed >= remainingTime) {
                throw new Error('Deadline exceeded');
            }

            const attemptTimeout = remainingTime - elapsed;
            
            try {
                const response = await this.makeSingleRequest(body, attemptTimeout, idempotencyKey);
                return JSON.parse(response.data);
            } catch (error) {
                lastError = error;
                
                if (error.message.includes('429')) {
                    // Rate limited - respect Retry-After
                    const retryAfter = parseInt(error.headers?.['retry-after'] || '1') * 1000;
                    if (elapsed + retryAfter >= remainingTime) {
                        throw new Error('Deadline exceeded');
                    }
                    await this.sleep(retryAfter);
                    continue;
                }
                
                if (error.message.includes('5xx') || error.message.includes('timeout')) {
                    // Retry on server errors and timeouts
                    const backoff = this.calculateBackoff(attempt);
                    if (elapsed + backoff >= remainingTime) {
                        throw new Error('Deadline exceeded');
                    }
                    await this.sleep(backoff);
                    continue;
                }
                
                // Don't retry on client errors (4xx except 429)
                throw error;
            }
        }

        throw lastError || new Error('Max retries exceeded');
    }

    async makeSingleRequest(body, timeout, idempotencyKey) {
        return new Promise((resolve, reject) => {
            const parsedUrl = new URL(this.config.baseUrl);
            const isHttps = parsedUrl.protocol === 'https:';
            const client = isHttps ? https : http;
            
            const options = {
                hostname: parsedUrl.hostname,
                port: parsedUrl.port || (isHttps ? 443 : 80),
                path: parsedUrl.pathname + '/v1/chat/completions',
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Content-Length': Buffer.byteLength(body),
                    'Idempotency-Key': idempotencyKey,
                    'Connection': 'keep-alive'
                },
                agent: this.agent,
                timeout: timeout
            };

            const req = client.request(options, (res) => {
                const buffer = this.bufferPool.get();
                let data = '';
                
                res.on('data', (chunk) => {
                    data += chunk.toString();
                });
                
                res.on('end', () => {
                    this.bufferPool.return(buffer);
                    
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve({ data, statusCode: res.statusCode });
                    } else if (res.statusCode === 429) {
                        reject(new Error(`429 Rate Limited`));
                    } else if (res.statusCode >= 500) {
                        reject(new Error(`5xx Server Error: ${res.statusCode}`));
                    } else {
                        reject(new Error(`Client Error: ${res.statusCode}`));
                    }
                });
            });

            req.on('error', (err) => {
                if (err.code === 'ECONNRESET' || err.code === 'ETIMEDOUT') {
                    reject(new Error('timeout'));
                } else {
                    reject(err);
                }
            });

            req.on('timeout', () => {
                req.destroy();
                reject(new Error('timeout'));
            });

            req.write(body);
            req.end();
        });
    }

    streamRequest(body, timeout, idempotencyKey) {
        const emitter = new EventEmitter();
        
        // Start the stream asynchronously
        setImmediate(async () => {
            try {
                await this.processStream(body, timeout, idempotencyKey, emitter);
            } catch (error) {
                emitter.emit('error', error);
            }
        });
        
        return emitter;
    }

    async processStream(body, timeout, idempotencyKey, emitter) {
        const parsedUrl = new URL(this.config.baseUrl);
        const isHttps = parsedUrl.protocol === 'https:';
        const client = isHttps ? https : http;
        
        const options = {
            hostname: parsedUrl.hostname,
            port: parsedUrl.port || (isHttps ? 443 : 80),
            path: parsedUrl.pathname + '/v1/chat/completions',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(body),
                'Idempotency-Key': idempotencyKey,
                'Connection': 'keep-alive'
            },
            agent: this.agent,
            timeout: timeout
        };

        const req = client.request(options, (res) => {
            if (!res.statusCode || res.statusCode < 200 || res.statusCode >= 300) {
                emitter.emit('error', new Error(`HTTP ${res.statusCode}`));
                return;
            }

            let buffer = this.bufferPool.get();
            let currentData = '';
            
            res.on('data', (chunk) => {
                buffer.write(chunk.toString());
                currentData += chunk.toString();
                
                // Parse SSE events with backpressure
                const events = this.parseSSEEvents(currentData);
                for (const event of events) {
                    if (event.data === '[DONE]') {
                        emitter.emit('end');
                        return;
                    }
                    
                    try {
                        const json = JSON.parse(event.data);
                        emitter.emit('data', json);
                    } catch (e) {
                        // Skip malformed JSON
                    }
                }
            });
            
            res.on('end', () => {
                this.bufferPool.return(buffer);
                emitter.emit('end');
            });
            
            res.on('error', (err) => {
                this.bufferPool.return(buffer);
                emitter.emit('error', err);
            });
        });

        req.on('error', (err) => {
            emitter.emit('error', err);
        });

        req.on('timeout', () => {
            req.destroy();
            emitter.emit('error', new Error('timeout'));
        });

        req.write(body);
        req.end();
    }

    parseSSEEvents(text) {
        const events = [];
        const lines = text.split('\n');
        let currentEvent = {};
        
        for (const line of lines) {
            if (line.startsWith('data: ')) {
                currentEvent.data = line.substring(6);
                events.push(currentEvent);
                currentEvent = {};
            }
        }
        
        return events;
    }

    calculateBackoff(attempt) {
        const delay = this.config.retryBaseDelay * Math.pow(2, attempt - 1);
        const jitter = delay * 0.1 * Math.random();
        return Math.min(delay + jitter, this.config.maxRetryDelay);
    }

    generateIdempotencyKey() {
        return `node-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

module.exports = { PerformanceClient, Semaphore, BufferPool };

File: python/src/performance_client.py
"""
# High-Performance Python LLM Client

Addresses all performance failure modes:
- Connection pooling with keep-alive
- Single deadline propagated to all operations
- Proper streaming with backpressure
- Bounded concurrency with asyncio.Semaphore
- Memory-efficient buffer reuse
"""

import asyncio
import json
import time
import uuid
from typing import Dict, List, Optional, AsyncGenerator, Any
from dataclasses import dataclass
import aiohttp
from aiohttp import ClientSession, ClientTimeout, TCPConnector


@dataclass
class ClientConfig:
    base_url: str = "http://localhost:3000"
    timeout: float = 30.0
    max_concurrent: int = 32
    keep_alive: float = 60.0
    retry_attempts: int = 3
    retry_base_delay: float = 0.1
    max_retry_delay: float = 5.0


class BufferPool:
    """Memory-efficient buffer pool to avoid allocations"""
    
    def __init__(self, max_size: int = 10):
        self.pool: List[bytearray] = []
        self.max_size = max_size
    
    def get(self) -> bytearray:
        if self.pool:
            buffer = self.pool.pop()
            buffer.clear()
            return buffer
        return bytearray(8192)
    
    def return_buffer(self, buffer: bytearray) -> None:
        if len(buffer) <= 65536 and len(self.pool) < self.max_size:
            buffer.clear()
            self.pool.append(buffer)


class PerformanceClient:
    """High-performance LLM client with all optimizations"""
    
    def __init__(self, config: Optional[ClientConfig] = None):
        self.config = config or ClientConfig()
        self.semaphore = asyncio.Semaphore(self.config.max_concurrent)
        self.buffer_pool = BufferPool()
        self._session: Optional[ClientSession] = None
    
    async def __aenter__(self):
        await self._ensure_session()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._session:
            await self._session.close()
    
    async def _ensure_session(self) -> None:
        """Create optimized HTTP session with connection pooling"""
        if self._session is None or self._session.closed:
            # Connection pooling with keep-alive
            connector = TCPConnector(
                limit=self.config.max_concurrent,
                limit_per_host=self.config.max_concurrent,
                keepalive_timeout=self.config.keep_alive,
                enable_cleanup_closed=True,
                ttl_dns_cache=300,  # DNS caching
                use_dns_cache=True,
            )
            
            timeout = ClientTimeout(
                total=self.config.timeout,
                connect=5.0,
                sock_read=self.config.timeout
            )
            
            self._session = ClientSession(
                connector=connector,
                timeout=timeout,
                headers={'Connection': 'keep-alive'}
            )
    
    async def chat_completion(
        self, 
        messages: List[Dict[str, str]], 
        deadline: float
    ) -> Dict[str, Any]:
        """Make chat completion request with deadline and concurrency control"""
        async with self.semaphore:  # Bound concurrency
            await self._ensure_session()
            
            # Check deadline
            if time.time() > deadline:
                raise asyncio.TimeoutError("Deadline exceeded")
            
            remaining_time = deadline - time.time()
            idempotency_key = self._generate_idempotency_key()
            
            body = {
                "model": "test-model",
                "messages": messages,
                "max_tokens": 100
            }
            
            return await self._make_request_with_retries(body, remaining_time, idempotency_key)
    
    async def stream_chat_completion(
        self, 
        messages: List[Dict[str, str]], 
        deadline: float
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """Stream chat completion with backpressure control"""
        async with self.semaphore:
            await self._ensure_session()
            
            if time.time() > deadline:
                raise asyncio.TimeoutError("Deadline exceeded")
            
            remaining_time = deadline - time.time()
            idempotency_key = self._generate_idempotency_key()
            
            body = {
                "model": "test-model",
                "messages": messages,
                "max_tokens": 100,
                "stream": True
            }
            
            async for chunk in self._stream_request(body, remaining_time, idempotency_key):
                yield chunk
    
    async def _make_request_with_retries(
        self, 
        body: Dict[str, Any], 
        remaining_time: float, 
        idempotency_key: str
    ) -> Dict[str, Any]:
        """Make request with retries and deadline enforcement"""
        attempt = 0
        last_error = None
        start_time = time.time()
        
        while attempt < self.config.retry_attempts:
            attempt += 1
            
            elapsed = time.time() - start_time
            if elapsed >= remaining_time:
                raise asyncio.TimeoutError("Deadline exceeded")
            
            attempt_timeout = remaining_time - elapsed
            
            try:
                response_data = await self._make_single_request(body, attempt_timeout, idempotency_key)
                return json.loads(response_data)
            except Exception as error:
                last_error = error
                
                if "429" in str(error):
                    # Rate limited - respect Retry-After
                    retry_after = 1.0  # Default
                    if hasattr(error, 'headers') and 'retry-after' in error.headers:
                        retry_after = float(error.headers['retry-after'])
                    
                    if elapsed + retry_after >= remaining_time:
                        raise asyncio.TimeoutError("Deadline exceeded")
                    
                    await asyncio.sleep(retry_after)
                    continue
                
                if "5xx" in str(error) or isinstance(error, asyncio.TimeoutError):
                    # Retry on server errors and timeouts
                    backoff = self._calculate_backoff(attempt)
                    if elapsed + backoff >= remaining_time:
                        raise asyncio.TimeoutError("Deadline exceeded")
                    
                    await asyncio.sleep(backoff)
                    continue
                
                # Don't retry on client errors (4xx except 429)
                raise error
        
        raise last_error or Exception("Max retries exceeded")
    
    async def _make_single_request(
        self, 
        body: Dict[str, Any], 
        timeout: float, 
        idempotency_key: str
    ) -> str:
        """Make single HTTP request with timeout"""
        url = f"{self.config.base_url}/v1/chat/completions"
        
        headers = {
            'Content-Type': 'application/json',
            'Idempotency-Key': idempotency_key
        }
        
        try:
            async with self._session.post(
                url, 
                json=body, 
                headers=headers,
                timeout=aiohttp.ClientTimeout(total=timeout)
            ) as response:
                if response.status >= 200 and response.status < 300:
                    return await response.text()
                elif response.status == 429:
                    raise Exception(f"429 Rate Limited")
                elif response.status >= 500:
                    raise Exception(f"5xx Server Error: {response.status}")
                else:
                    raise Exception(f"Client Error: {response.status}")
        except asyncio.TimeoutError:
            raise asyncio.TimeoutError("Request timeout")
        except Exception as e:
            raise Exception(f"Request failed: {e}")
    
    async def _stream_request(
        self, 
        body: Dict[str, Any], 
        timeout: float, 
        idempotency_key: str
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """Stream request with backpressure control"""
        url = f"{self.config.base_url}/v1/chat/completions"
        
        headers = {
            'Content-Type': 'application/json',
            'Idempotency-Key': idempotency_key
        }
        
        try:
            async with self._session.post(
                url, 
                json=body, 
                headers=headers,
                timeout=aiohttp.ClientTimeout(total=timeout)
            ) as response:
                if response.status < 200 or response.status >= 300:
                    raise Exception(f"HTTP {response.status}")
                
                buffer = self.buffer_pool.get()
                current_data = ''
                
                try:
                    async for chunk in response.content.iter_chunked(1024):
                        buffer.extend(chunk)
                        current_data += chunk.decode('utf-8', errors='ignore')
                        
                        # Parse SSE events with backpressure
                        events = self._parse_sse_events(current_data)
                        for event in events:
                            if event.get('data') == '[DONE]':
                                return
                            
                            try:
                                json_data = json.loads(event['data'])
                                yield json_data
                            except json.JSONDecodeError:
                                # Skip malformed JSON
                                pass
                finally:
                    self.buffer_pool.return_buffer(buffer)
                    
        except asyncio.TimeoutError:
            raise asyncio.TimeoutError("Stream timeout")
        except Exception as e:
            raise Exception(f"Stream failed: {e}")
    
    def _parse_sse_events(self, text: str) -> List[Dict[str, str]]:
        """Parse Server-Sent Events efficiently"""
        events = []
        lines = text.split('\n')
        current_event = {}
        
        for line in lines:
            if line.startswith('data: '):
                current_event['data'] = line[6:]
                events.append(current_event)
                current_event = {}
        
        return events
    
    def _calculate_backoff(self, attempt: int) -> float:
        """Calculate exponential backoff with jitter"""
        delay = self.config.retry_base_delay * (2 ** (attempt - 1))
        jitter = delay * 0.1 * (0.5 - 0.5)  # Simplified jitter
        return min(delay + jitter, self.config.max_retry_delay)
    
    def _generate_idempotency_key(self) -> str:
        """Generate unique idempotency key"""
        return f"py-{int(time.time() * 1000)}-{uuid.uuid4().hex[:9]}"


# Example usage and test
async def smoke_test():
    """Quick smoke test of the performance client"""
    config = ClientConfig()
    
    async with PerformanceClient(config) as client:
        messages = [{"role": "user", "content": "Hello"}]
        deadline = time.time() + 10.0  # 10 second deadline
        
        try:
            # Test regular completion
            result = await client.chat_completion(messages, deadline)
            print(f"✅ Chat completion: {result.get('id', 'unknown')}")
            
            # Test streaming
            print("🧪 Testing streaming...")
            async for chunk in client.stream_chat_completion(messages, deadline):
                print(f"📦 Chunk: {chunk}")
                
        except Exception as e:
            print(f"❌ Test failed: {e}")


if __name__ == "__main__":
    asyncio.run(smoke_test())




